feat(mention): add mention for commands

cptpiepmatz opened the pull request:
I realized that the `mention` crate is missing `Display` implementations for commands. Commands are formatted like this:
- `</NAME:COMMAND_ID>` for commands
- `</NAME SUBCOMMAND:ID>` for subcommands
- `</NAME SUBCOMMAND_GROUP SUBCOMMAND:ID>` for subcommand groups

This is documented [here](https://discord.com/developers/docs/reference#message-formatting) and in the [changelog](https://discord.com/developers/docs/change-log#slash-command-mentions).

For the `Mention` implementation I checked what could be useful implementations. The tuples here are for convenience but maybe obsolete.

I considered implementing `Mention` for `model::application::command::Command` and `model::application::application_command::CommandData` but both them may contain mutiple subcommands or subcommand groups, therefore they are not feasible to implement.

Also the implementation needs not only the command id but also name (and maybe also subcommand and subcommand group). That's why I added the new enum. Three structs could also solve this but this was my personal preference.

laralove143 commented on twilight-mention/src/fmt.rs:
```diff
@@ -118,6 +166,51 @@ impl Mention<Id<ChannelMarker>> for Id<ChannelMarker> {
 }
 
+/// Mention a command. This will format as `</NAME:COMMAND_ID>`.
+impl<N> Mention<CommandMention> for (N, Id<CommandMarker>)
```
this could be a struct instead of a tuple for clarity

cptpiepmatz replied:
removed

laralove143 commented on twilight-mention/src/fmt.rs:
```diff
@@ -118,6 +166,51 @@ impl Mention<Id<ChannelMarker>> for Id<ChannelMarker> {
+}
+
+/// Mention a subcommand. This will format as `</NAME SUBCOMMAND:ID>`.
+impl<N, S> Mention<CommandMention> for (N, S, Id<CommandMarker>)
```
likewise, this could be a struct

cptpiepmatz replied:
removed

laralove143 commented on twilight-mention/src/fmt.rs:
```diff
@@ -118,6 +166,51 @@ impl Mention<Id<ChannelMarker>> for Id<ChannelMarker> {
+}
+
+/// Mention a subcommand group. This will format as `</NAME SUBCOMMAND_GROUP SUBCOMMAND:ID>`.
+impl<N, G, S> Mention<CommandMention> for (N, G, S, Id<CommandMarker>)
```
same as above, this could be a struct

cptpiepmatz replied:
removed

laralove143 commented on twilight-mention/src/fmt.rs:
```diff
@@ -189,13 +282,45 @@ impl Mention<Id<UserMarker>> for User {
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum CommandMention {
+    // Field order is the same as the mention format.
+    Command {
```
turning the tuples into structs, these should wrap over the structs, or the implementations could be removed and users could rely on this enum entirely

cptpiepmatz replied:
removed tuples, more structs would be bloat

laralove143 commented on twilight-mention/src/fmt.rs:
```diff
@@ -189,13 +282,45 @@ impl Mention<Id<UserMarker>> for User {
+#[allow(missing_docs)]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum CommandMention {
+    // Field order is the same as the mention format.
```
i think we should still order fields alphabetically like we do (almost) everywhere

cptpiepmatz replied:
I ordered them now.

laralove143 reviewed (changes requested):
thank you for this catch

cptpiepmatz commented:
I'm personally not quite sure whether to go with three structs and put them into the enum or just remove the implementations on the tuples.

I think three more structs would be a too much bloat, so I'm considering simply removing the tuple impls. What do you think @laralove143?

laralove143 commented:
i agree, the enum and the impl on command are probably enough

cptpiepmatz commented:
Ok, I removed the tuple implementations and ordered the fields of the enum variants.

cptpiepmatz commented:
While fixing the tests I realized that you have to wrap `CommandMention` inside `MentionFormat` manually, should I make a `Mention` implementation for `CommandMention` or directly implement `Display` for `CommandMention`?

@laralove143

laralove143 commented:
> While fixing the tests I realized that you have to wrap `CommandMention` inside `MentionFormat` manually, should I make a `Mention` implementation for `CommandMention` or directly implement `Display` for `CommandMention`?
> 
> @laralove143

yes please, since thats how it works for other types

also we should have `ParseMention` for it as well

cptpiepmatz commented:
> yes please, since thats how it works for other types

I guess you mean to implement `Mention` for `CommandMention`, right?

cptpiepmatz commented:
@laralove143 The impl for `Mention` for `CommandMention` would need a clone:
```rs
impl Mention<CommandMention> for CommandMention {
    fn mention(&self) -> MentionFormat<CommandMention> {
        MentionFormat(self.clone())
    }
}
```

Should we keep that?

laralove143 commented:
> @laralove143 The impl for `Mention` for `CommandMention` would need a clone:
> 
> ```rust
> impl Mention<CommandMention> for CommandMention {
>     fn mention(&self) -> MentionFormat<CommandMention> {
>         MentionFormat(self.clone())
>     }
> }
> ```
> 
> Should we keep that?

yes, as the alternative is to add a method to `Mention` that consumes `self` or take a `&CommandMention`, i think cloning small strings is negligible

cptpiepmatz commented:
> yes, as the alternative is to add a method to `Mention` that consumes self or take a `&CommandMention`, i think cloning small strings is negligible

I ended up doing a bit of both. This way, users can avoid the cloning if they prefer. The inner field of `MentionFormat` is private, so we needed to expose something to allow for this. I implemented `Mention` for `CommandMention` and noted in the documentation that this method uses `clone`. I also added the `into_mention` function that takes `self` and creates a `MentionFormat<CommandMention>` without any cloning.

> also we should have `ParseMention` for it as well

I took care of this in commit 808b1c0720a958717534848b1265a7949d2d9eb9. I also added a new variant to `ParseMentionErrorType` for cases where too many segments are found. The `parse_mention` method didn't really fit because it has a different pattern, and I didn't want to alter it, so the `parse` method is a bit longer than usual.

By the way, I saw that we have a `MentionType` enum. I was considering adding `CommandMention` to it, but that would mean it's no longer `Copy`. What do you think we should do here, @laralove143? Making `CommandMention` `Copy` doesn't feel right to me, as it's not as simple as cloning an `Id`.

laralove143 commented on twilight-mention/src/parse/impl.rs:
```diff
@@ -84,6 +86,132 @@ impl ParseMention for Id<ChannelMarker> {
 }
 
+impl ParseMention for CommandMention {
+    const SIGILS: &'static [&'static str] = &["/"];
```
i dont think we have to provide the lifetime explicitly here

cptpiepmatz replied:
Other implementations do this too.

laralove143 commented on twilight-mention/src/parse/impl.rs:
```diff
@@ -201,6 +329,52 @@ impl ParseMention for Id<UserMarker> {
+            })
+        }
+        (Some(name), None, None, None) => Ok(CommandMention::Command {
+            name: name.to_string(),
```
we should use `.to_owned` here, as thats what twilight generally uses

cptpiepmatz replied:
done

laralove143 reviewed (changes requested):
thank you for the changes, theres only a few nits remaining

cptpiepmatz commented:
> about `MentionType`, it wouldnt make a difference in practice, it'd still use `Copy` for the other variants, so i think its okay to add the new variant there

You cannot derive `Copy` as `CommandMention` isn't `Copy`. And I don't think that `CommandMention` should be `Copy`. I would be fine to manually implement `Copy` for a `MentionType` with a `Command` variant.

laralove143 commented:
> > about `MentionType`, it wouldnt make a difference in practice, it'd still use `Copy` for the other variants, so i think its okay to add the new variant there
> 
> You cannot derive `Copy` as `CommandMention` isn't `Copy`. And I don't think that `CommandMention` should be `Copy`. I would be fine to manually implement `Copy` for a `MentionType` with a `Command` variant.

i think marking it `Copy` when it _might_ not be copy is misleading but taking away `Copy` from `MentionType` would make this pr breaking, so i think we should open another pr for adding `CommandMention` to `MentionType` that targets `next`

cptpiepmatz commented:
That sounds like the best approach to me. So is this done?

laralove143 reviewed (approved):
thank you for the addition!

AEnterprise reviewed (commented):
looks good, just a small issue left in the docs build with a redundant import

cptpiepmatz commented:
@AEnterprise I checked the logs of that failing build, it fails on:
> twilight-http/src/json.rs:4:29

I never touched that file in my PR.

laralove143 commented:
im sure `JsonDeserializer` is actually used, maybe not in conditional compiling (used under `#[cfg(not(doc))]`) but this is very unlikely, i think its a false positive, maybe because of incremental building (it is used but not in the changes in this pr) ill go ahead and merge this
